pusher.go

```diff
@@ -0,0 +1,298 @@
+	return p.authenticate(&authentications.PresenceChannel{Body: body, Member: member})
+}
+
+func (p *Pusher) authenticate(request authentications.Request) (response []byte, err error) {
+	var unsigned string
+	if unsigned, err = request.StringToSign(); err != nil {
+		return
+	}
+	authSignature := signatures.HMAC(unsigned, p.secret)
+
+	responseMap := map[string]string{
+		"auth": fmt.Sprintf("%s:%s", p.key, authSignature),
+	}
+	var userData string
+	if userData, err = request.UserData(); err != nil {
+		return
+	}
+	if userData != "" {
+		responseMap["channel_data"] = userData
+	}
+	return json.Marshal(responseMap)
+}
+
+func (p *Pusher) Notify(interests []string, notification *Notification) (response []NotificationResponse, err error) {
+	if len(interests) != 1 {
+		err = errors.New("Currently only sending to one interest is supported.")
+		return
+	}
+
+	if notification.Gcm != nil {
+		ttl := notification.Gcm.TimeToLive
+		if ttl != nil && (*ttl < uint(0) || uint(2419200) < *ttl) {
+			err = errors.New("GCM's TimeToLive field must be an integer between 0 and 2419200 (4 weeks)")
+			return
+		}
+
+		if notification.Gcm.Payload != nil {
+			payload := notification.Gcm.Payload
+			if len(payload.Title) == 0 {
+				err = errors.New("Notification title is a required field for GCM")
+				return
+			}
+
+			if len(payload.Icon) == 0 {
+				err = errors.New("Notification icon is a required field for GCM")
+				return
+			}
+
+		}
+
+		if notification.WebhookURL != "" {
+			if _, err = url.Parse(notification.WebhookURL); err != nil {
+				return
+			}
+		}
+	}
+
+	if !map[string]bool{"INFO": true, "DEBUG": true, "": true}[notification.WebhookLevel] {
+		err = errors.New("Webhook level must be either INFO or DEBUG. Blank will default to INFO")
+		return
+	}
+
+	var body, byteResponse []byte
+
+	req := &notificationRequest{
+		Interests:    interests,
+		Notification: notification,
+	}
+
+	config := &urlConfig{
+		key:    p.key,
+		secret: p.secret,
+		host:   NATIVE_NOTIFICATIONS_HOST,
+		appID:  p.appID,
+	}
+
+	if body, err = json.Marshal(&req); err != nil {
+		return
+	}
+
+	params := &requests.Params{
+		Body: body,
+	}
+
+	if byteResponse, err = p.sendRequest(config, p.httpClient(), requests.NativePush, params); err != nil {
+		return
+	}
+
+	err = json.Unmarshal(byteResponse, &response)
+	return
+}
+
+func (p *Pusher) Webhook(header http.Header, body []byte) (webhook *Webhook, err error) {
+	for _, token := range header["X-Pusher-Key"] {
+		if token == p.key && signatures.CheckHMAC(header.Get("X-Pusher-Signature"), p.secret, body) {
+			return newWebhook(body)
+		}
+	}
+	err = errors.New("Invalid webhook")
+	return
+}
+
+func (p *Pusher) urlConfig() *urlConfig {
+	return &urlConfig{
+		key:     p.key,
+		secret:  p.secret,
+		host:    p.Host,
+		cluster: p.Cluster,
+		appID:   p.appID,
+		secure:  p.Secure,
+	}
+}
+
+func (p *Pusher) httpClient() *http.Client {
+	if p.HttpClient == nil {
+		p.HttpClient = &http.Client{Timeout: time.Second * 5}
```

Comment:
I think this timeout should be configurable by the library's clients, at least that's the case for the other languages.